samcli/lib/build/app_builder.py

```diff
@@ -0,0 +1,137 @@
+            "manifest_name": "requirements.txt"
+        },
+        "python3.6": {
+           "language": "python",
+           "dependency_manager": "pip",
+           "application_framework": None,
+           "manifest_name": "requirements.txt"
+        }
+    }
+
+    def __init__(self, function_provider, build_dir, source_root, use_container=False, parallel=False):
+        """
+        Initialize the class
+
+        Parameters
+        ----------
+        function_provider : samcli.commands.local.lib.sam_function_provider.SamFunctionProvider
+            Provider that can vend out functions available in the SAM template
+
+        build_dir : str
+            Path to the directory where we will be storing built artifacts
+
+        source_root : str
+            Path to a folder. Use this folder as the root to resolve relative source code paths against
+
+        use_container : bool
+            Optional. Set to True if you want to run the builds on an container that simulates AWS Lambda environment
+
+        parallel : bool
+            Optional. Set to True to build each function in parallel to improve performance
+        """
+        self.function_provider = function_provider
+        self.build_dir = build_dir
+        self.source_root = source_root
+
+        self.use_container = use_container
+        self.parallel = parallel
+
+    def build(self):
+        """
+        Build the entire application
+
+        Returns
+        -------
+        dict
+            Returns the path to where each resource was built as a map of resource's LogicalId to the path string
+        """
+
+        result = {}
+
+        for lambda_function in self.function_provider.get_all():
+            result[lambda_function.name] = self._build_function(lambda_function.name,
+                                                                lambda_function.codeuri,
+                                                                lambda_function.runtime)
+
+        return result
+
+    def update_template(self, template_dict, built_artifacts):
+        """
+        Given the path to built artifacts, update the template to point appropriate resource CodeUris to the artifacts
+        folder
+
+        Parameters
+        ----------
+        template_dict
+        built_artifacts : dict
+            Map of LogicalId of a resource to the path where the the built artifacts for this resource lives
+
+        Returns
+        -------
+        dict
+            Updated template
+        """
+        # TODO: Move this method to a "build provider" or some other class. It doesn't really fit within here.
+
+        for logical_id, resource in template_dict.get("Resources", {}).items():
+
+            if logical_id not in built_artifacts:
+                # this resource was not built. So skip it
+                continue
+
+            resource_type = resource.get("Type")
+            if resource_type == "AWS::Serverless::Function":
```

Comment:
We need a better way to do these things.

Reply:
Dictionaries for code branching, makes code look cleaner as well :)
store functions as values to a dictionary key "AWS::Serverless::Function", "AWS::Lambda::Function"